### PR TITLE
[android] Add Hermes support for Expo Go

### PR DESCRIPTION
**Pull request by Kudo**
# Why

To support loading hermes bytecode bundle from Expo Go.
Close [ENG-1276](https://linear.app/expo/issue/ENG-1276)

# How

We pack both libjsc.so and libhermes.so into Expo Go.
And create a JavaScriptExecutorFactory that will do the following things:
- If the loading bundle is a hermes bytecode bundle and the bytecode version is supported, use the `HermesExectorFactory`
- Otherwise, use the `JSCExecutorFactory`

Since there may be different hermes-engine version between Expo SDK upgrade, add libhermes.so into versioning scope.

If Expo Go loads an incompatible hermes bytecode version, will show an error on ErrorActivity.

## APK size analysis
Release build
Before: 71MB
After: 74MB

We could assume each SDK version would introduce additional 3~4MB APK size.

# Test Plan

### Unit Test

```
android/expoview/src/test/java/versioned/host/exp/exponent/VersionedUtilsTest.kt

isHermesBundle_plainJS_returnFalse
isHermesBundle_hermesBytecodeBundle74_returnTrue
isHermesBundle_nullInput_returnFalse
getHermesBundleBytecodeVersion_plainJS_return0
getHermesBundleBytecodeVersion_hermesBytecodeBundle74_return74
```

### Handy test

- [x] Launch unversioned Expo Go and load remote hermes bundle
- [x]  Launch unversioned Expo Go and load remote plaintext js bundle
- [x] `et add-sdk -p android -s 42.0.0`
- [x] Launch versioned Expo Go and load remote hermes bundle
- [x] Launch versioned Expo Go and load remote plaintext js bundle
- [x] Load remote NCL hermes bundle

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: android/expoview/src/test/java/versioned/host/exp/exponent/VersionedUtilsTest.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    assert(jsBundlePath != null)
+    assert(!VersionedUtils.isSupportedHermesBundle(jsBundlePath))
+  }
+
+  @Test
+  fun isSupportedHermesBundle_nullInput_returnFalse() {
+    assert(!VersionedUtils.isSupportedHermesBundle(null))
+    assert(!VersionedUtils.isSupportedHermesBundle(""))
+  }
+}
```
Comment (tsapeta) [inline android/expoview/src/test/java/versioned/host/exp/exponent/VersionedUtilsTest.kt:32]:
```suggestion
}

```

File: tools/src/versioning/android/index.ts [hunk 2]
```diff
@@ -608,6 +616,15 @@ async function exportReactNdksIfNeeded() {
   }
 }
 
+async function addBuildGradleRenameTaskForPrebuiltLibs(version: string) {
+  const abiVersion = version.replace(/\./g, '_');
+  const filePath = path.join(versionedReactAndroidPath, 'build.gradle');
+  let content = await fs.readFile(filePath, 'utf8');
+  const renameTask = `        rename '(.+).so', '$$1_abi${abiVersion}.so'\n`;
+  content = content.replace(/(into "\$thirdPartyNdkDir\/hermes"\n)(\s+?\})/gm, `$1${renameTask}$2`);
+  await fs.writeFile(filePath, content);
```
Comment (tsapeta) [inline tools/src/versioning/android/index.ts:625]: Try using `transformFileAsync` function from `Transforms.ts` — it would make this code cleaner 😉

File: android/expoview/src/test/java/versioned/host/exp/exponent/VersionedUtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  @Test
+  fun isSupportedHermesBundle_plainJS_returnFalse() {
+    val jsBundlePath = VersionedUtilsTest::class.java.getResource("/plain.js")?.path
+    assert(jsBundlePath != null)
+    assert(!VersionedUtils.isSupportedHermesBundle(jsBundlePath))
+  }
+
+  @Test
+  fun isSupportedHermesBundle_hermesBytecodeBundle74_returnTrue() {
+    val jsBundlePath = VersionedUtilsTest::class.java.getResource("/plain.74.hbc")?.path
```
[the remark this answers is not in the document]
Comment (Kudo) [reply android/expoview/src/test/java/versioned/host/exp/exponent/VersionedUtilsTest.kt:15]: https://github.com/expo/expo/pull/13238/commits/8818a92eac32e13b2241f3a717ab0d709445423d addressed this issue

**Comment by Kudo** [comment]
Some updates in summary:

1. APK size increased about 3MB (which assumed to increased for each Expo SDK in the future)
2. Checked versioned test and load remote hermes bundle. There were some versioning broken on master, so to fallback sdk-41 and verified on sdk41.
3. Test unversioned [NCL hermes bundle](https://expo.io/@kudochien/native-component-list), mostly ok. There were some crash, but seems to happened before this pr. e.g. Localization screen.
4. If loading a bundle with incompatible hermes bytecode version, will show error on ErrorActivity.

**Comment by brentvatne** [comment]
this looks good! pending @tsapeta's review i think we should move forward with this

**Review by brentvatne — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

**Comment by brentvatne** [comment]
let's land this to unblock versioning, but @tsapeta pointed out that it would be good to have @wkozyra95 review post-merge in case we missed anything

**Comment by wkozyra95** [comment]
> let's land this to unblock versioning, but @tsapeta pointed out that it would be good to have @wkozyra95 review post-merge in case we missed anything

LGTM

@Kudo One concern I have is whether it will work for the next sdk, if we switch between apps from different sdks wouldn't it cause loading the same hermes lib twice and fail with symbol collisions. 
I'm not sure if it's a problem or not, technically it works with react-native and reanimated libs, but in the case of expo-gl-cpp we are including one unversioned binary because of those symbols colisions

**Comment by lukmccall** [comment]
@Kudo did you check if Reanimated v2 works with Hermes? By default, this package is built without support for Hermes - (see https://github.com/expo/expo/blob/b032029e5c20834ce367e1f22c0ae984a47a1bc7/android/expoview/build.gradle#L72). I believe, we need to have two separate versions of reanimated to support both JSC and Hermes.

**Comment by Kudo** [comment]
@wkozyra95 that is a good catch. i just versioning libhermes.so on filename but not to symbol level.
symbol collisions is possible as you mentioned, let me try the case for loading two hermes engines and get you back.

@lukmccall yes, reanimated2 works. the interesting part is that even bundle is executed by hermes, reanimated worklet is still running by jsc. do you foresee if there are any potential problems for this?

**Comment by lukmccall** [comment]
Ok, so maybe this isn't a problem. However, I think it will be nice in the future to ensure that worklets are running on the same VM. The reanimated team might add something which can break our current setup. But for now, we can leave it as it is.

**Comment by Kudo** [comment]
@lukmccall agree you. i like the way reanimated 2.1.0 to create js runtim, unfortunately it broke hermes inspector. maybe in the future, the right approach is to propose a `JSRuntimeFactory` from react native upstream.

**Comment by brentvatne** [comment]
re: switching between apps with different versions - we currently allow multiple projects to be open within one instance of expo go, but this is not necessary and i think limiting to one running project is fine if that simplifies this problem. we've been wanting to do a refactor of the android client to remove support for running multiple projects (and cleanup all of the extra complexity that comes with that) but haven't got around to it yet

**Comment by Kudo** [comment]
@wkozyra95 lately update: after react-native 0.64 upgrade is ready, i am able to verify expo-go hermes versioning easier. as expect, it crashes 😅
TL;DR: fix proposing in #13918 

## Test Plan (Copied from #13680)

based on #13793 
 - load [sdk42 hermes bundle](https://expo.dev/@kudochien/sdk42) which hbc version is 74
 - load local unversioned hermes project based on react-native 0.64 (hermes-engine 0.7.2) which hbc version is 76

expo-go will crash after loading these two projects without closing existing activities.

to load the sdk42 project first then the unversioned project, the stacktrace is something like that:

```
08-05 01:00:15.494  7479  7479 F DEBUG   : backtrace:
08-05 01:00:15.494  7479  7479 F DEBUG   :       #00 pc 0000000000036068  /data/app/~~WU4kibtVi4PluLyeAnc2Lw==/host.exp.exponent-GT5vlI_tvAnBIZklR462UA==/lib/arm64/libhermes_abi42_0_0.so (BuildId: 104108701ea35e3b32f33afab7ef38388a9e99aa)
08-05 01:00:15.494  7479  7479 F DEBUG   :       #01 pc 000000000003564c  /data/app/~~WU4kibtVi4PluLyeAnc2Lw==/host.exp.exponent-GT5vlI_tvAnBIZklR462UA==/lib/arm64/libhermes_abi42_0_0.so (facebook::hermes::HermesRuntimeImpl::HermesRuntimeImpl(hermes::vm::RuntimeConfig co
nst&)+140) (BuildId: 104108701ea35e3b32f33afab7ef38388a9e99aa)
08-05 01:00:15.494  7479  7479 F DEBUG   :       #02 pc 000000000002eb58  /data/app/~~WU4kibtVi4PluLyeAnc2Lw==/host.exp.exponent-GT5vlI_tvAnBIZklR462UA==/lib/arm64/libhermes_abi42_0_0.so (facebook::hermes::makeHermesRuntime(hermes::vm::RuntimeConfig const&)+40) (BuildId: 104108701ea35e3b32f33afab7ef38388a9e99aa)
08-05 01:00:15.494  7479  7479 F DEBUG   :       #03 pc 0000000000019090  /data/app/~~WU4kibtVi4PluLyeAnc2Lw==/host.exp.exponent-GT5vlI_tvAnBIZklR462UA==/lib/arm64/libhermes-executor-common-debug.so (facebook::react::HermesExecutorFactory::createJSExecutor(std::__ndk1::shared_ptr<facebook::react::ExecutorDelegate>, std::__ndk1::shared_ptr<facebook::react::MessageQueueThread>)+68) (BuildId: c483547cfb7ec4f607fb07422a0930d20c5dcf5b)
08-05 01:00:15.494  7479  7479 F DEBUG   :       #04 pc 00000000000a47f0  /data/app/~~WU4kibtVi4PluLyeAnc2Lw==/host.exp.exponent-GT5vlI_tvAnBIZklR462UA==/lib/arm64/libreactnativejni.so (facebook::react::NativeToJsBridge::NativeToJsBridge(facebook::react::JSExecutorFactory*, std::__ndk1::shared_ptr<facebook::react::ModuleRegistry>, std::__ndk1::shared_ptr<facebook::react::MessageQueueThread>, std::__ndk1::shared_ptr<facebook::react::InstanceCallback>)+236) (BuildId: 001cd77af95213519b49ac75f78b862bcf18b30f)
08-05 01:00:15.494  7479  7479 F DEBUG   :       #05 pc 000000000009ca88  /data/app/~~WU4kibtVi4PluLyeAnc2Lw==/host.exp.exponent-GT5vlI_tvAnBIZklR462UA==/lib/arm64/libreactnativejni.so (std::__ndk1::shared_ptr<facebook::react::NativeToJsBridge> std::__ndk1::shared_ptr<facebook::react::NativeToJsBridge>::make_shared<facebook::react::JSExecutorFactory*, std::__ndk1::shared_ptr<facebook::react::ModuleRegistry>&, std::__ndk1::shared_ptr<facebook::react::MessageQueueThread>&, std::__ndk1::shared_ptr<facebook::react::InstanceCallback>&>(facebook::react::JSExecutorFactory*&&, std::__ndk1::shared_ptr<facebook::react::ModuleRegistry>&, std::__ndk1::shared_ptr<facebook::react::MessageQueueThread>&, std::__ndk1::shared_ptr<facebook::react::InstanceCallback>&)+236) (BuildId: 001cd77af95213519b49ac75f78b862bcf18b30f)
```

unversioned libhermes-executor accidentally load the libhermes_abi_42_0_0.so.

## Research for solutions

### `objcopy --prefix-symbols`

my first thought is trying to add abi prefix to symbols. unfortunately, `object --prefix-symbols` cannot manipulate dynamic symbols.

### building hermes for source

that is supposedly the most reliable solution. however, building hermes takes about 30m for android and 60m for ios. we might better to publish prebuilt versioned libraries 😣

### `SONAME` did matter

that is the final proposing solution. i was curious that android did not specify `RTLD_GLOBAL` in [dlopen](https://android.googlesource.com/platform/art/+/1d9d3587dafb7078082b4d4058eea753c13ef94f/libnativeloader/native_loader.cpp#224), and it is [RTLD_LOCAL by default](https://android.googlesource.com/platform/bionic/+/android-n-mr2-preview-1/android-changes-for-ndk-developers.md#rtld_local-available-in-api-level-23). supposedly, the symbols should not pollute globally.

a guess is the `SONAME`. i did not update libhermes.so's soname from versioning. from objdump, the `DT_NEEDED` is still libhermes.so. my guess is that linker resolves libraries by soname. if a soname is already resolved, even if filename is different, linker will still locate symbols from the loaded & matched soname libraries.

thankfully, [patchelf](https://github.com/NixOS/patchelf) rescues for updating soname in binary level.
after versioning the soname, expo-go could now load two different hermes versions.
i am not pretty sure if there will be any other issues, at least expo-go could run. if there are any other issues, we could then turn to build hermes from source or force restart expo-go if hermes version changed.

## iOS versioning

Let's test this after next react-native upgrade 😅
